lib/pbench/test/functional/agent/cli/commands/tools/test_clear_tools.py

```diff
@@ -165,48 +199,200 @@ def test_clear_tools_test69(monkeypatch, agent_config, pbench_run, pbench_cfg):
+            # of the remote is not empty
+            elif tool.parent.parent.exists():
+                for remote_dir in tool.parent.parent.iterdir():
+                    assert any(
+                        remote_dir.iterdir()
+                    ), f'Found an empty "remote" directory, "{remote_dir!s}", which should have been deleted.'
+        else:
+            assert tool.exists()
+
+
+@pytest.mark.parametrize(
+    "groups", ["", "default,wrong_group", "group1,wrong_group"],
+)
+@pytest.mark.parametrize(
+    "remotes", ["", "remote1,wrong_remote1", "wrong_remote1,wrong_remote2"],
+)
+@pytest.mark.parametrize(
+    "tools", ["", "pidstat,wrong_tool1", "wrong_tool1,wrong_tool2"],
+)
+def test_clear_tools_test86(tools_configuration, groups, remotes, tools):
+    """ Attempt to clear tools at various group-remote-tool locations where the
+    combination may not present on the hosts"""
+    _, tools_tree = tools_configuration
+    command = ["pbench-clear-tools"]
+    if groups:
+        command.append(f"--groups={groups}")
+    if remotes:
+        command.append(
+            f"--remotes={','.join(map('{}.example.com'.format, remotes.split(',')))}"
+        )
+    if tools:
+        command.append(f"--name={tools}")
+
+    out, err, exitcode = pytest.helpers.capture(command)
+    assert b"" == out
+
+    if not groups:
+        groups = "default"
+
+    for group in groups.split(","):
+        err_msg = f'pbench-clear-tools: invalid --group option "{group}"'
+        if group not in tools_tree.keys():
+            assert err_msg.encode() in err
+        else:
+            assert err_msg.encode() not in err
+            for remote in remotes.split(","):
+                err_msg = f'No remote host "{remote}.example.com"'
+                if remote and remote not in tools_tree[group]:
+                    assert err_msg.encode() in err
+                elif remote:
+                    wrong_tools = []
+                    if tools:
+                        for tool in tools.split(","):
+                            if tool not in tools_tree[group][remote]:
+                                wrong_tools.append(tool)
+                    if wrong_tools:
+                        wrong_tools = sorted(wrong_tools)
+                        err_msg = f"Tools {wrong_tools} not found in remote"
+                        assert err_msg.encode() in err
+                    else:
+                        err_msg = f'All tools removed from group "{group}" on host "{remote}.example.com"'
+                        assert err_msg.encode() in err
+                else:
+                    remotes = tools_tree[group].keys()
+                    for remote in remotes:
+                        tools_not_found = list(
+                            set(tools.split(",")) - set(tools_tree[group][remote])
+                        )
+                        tools_not_found = [tool for tool in tools_not_found if tool]
```

Comment:
You don't _need_ to invoke the `list()` constructor at line 390 -- you can just as easily iterate over the `set`, instead, at line 393.  But, it turns out that, if you omit the `list()`, Black makes the code uglier, so I vote that you keep it.  😝